Mathlib/FieldTheory/IsSepClosed.lean

```diff
@@ -0,0 +1,174 @@
+/-
+Copyright (c) 2020 Kenny Lau, 2023 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Jz Pan
+-/
+import Mathlib.FieldTheory.IsAlgClosed.AlgebraicClosure
+
+/-!
+# Separably Closed Field
+
+In this file we define the typeclass for separably closed fields and separable closures,
+and prove some of their properties.
+
+## Main Definitions
+
+- `IsSepClosed k` is the typeclass saying `k` is a separably closed field, i.e. every separable
+polynomial in `k` splits.
+
+- `IsSepClosure R K` is the typeclass saying `K` is a separable closure of `R`, where `R` is a
+  commutative ring. This means that the map from `R` to `K` is injective, and `K` is
+  separably closed and separable over `R`
+
+## Tags
+
+separable closure, separably closed
+
+-/
+
+universe u v w
+
+open scoped Classical BigOperators Polynomial
+
+open Polynomial
+
+variable (k : Type u) [Field k]
+
+/-- Typeclass for separably closed fields.
+
+To show `Polynomial.Splits p f` for an arbitrary ring homomorphism `f`,
+see `IsSepClosed.splits_codomain` and `IsSepClosed.splits_domain`.
+-/
+class IsSepClosed : Prop where
+  splits_of_separable : ∀ p : k[X], p.Separable → (p.Splits <| RingHom.id k)
+
+/-- Every separable polynomial splits in the field extension `f : K →+* k` if `k` is
+separably closed.
+
+See also `IsSepClosed.splits_domain` for the case where `K` is separably closed.
+-/
+theorem IsSepClosed.splits_codomain {k K : Type _} [Field k] [IsSepClosed k] [Field K] {f : K →+* k}
+    (p : K[X]) (h : p.Separable) : p.Splits f := by
+  convert IsSepClosed.splits_of_separable (p.map f) (Separable.map h); simp [splits_map_iff]
+
+/-- Every separable polynomial splits in the field extension `f : K →+* k` if `K` is
+separably closed.
+
+See also `IsSepClosed.splits_codomain` for the case where `k` is separably closed.
+-/
+theorem IsSepClosed.splits_domain {k K : Type _} [Field k] [IsSepClosed k] [Field K] {f : k →+* K}
+    (p : k[X]) (h : p.Separable) : p.Splits f :=
+  Polynomial.splits_of_splits_id _ <| IsSepClosed.splits_of_separable _ h
+
+namespace IsSepClosed
+
+variable {k}
+
+theorem exists_root [IsSepClosed k] (p : k[X]) (hp : p.degree ≠ 0) (hsep : p.Separable) :
+    ∃ x, IsRoot p x :=
+  exists_root_of_splits _ (IsSepClosed.splits_of_separable p hsep) hp
+
+theorem exists_pow_nat_eq [IsSepClosed k] (x : k) {n : ℕ} (hn : (n : k) ≠ 0) : ∃ z, z ^ n = x := by
+  have hn' : 0 < n := Nat.pos_of_ne_zero <| fun h => by
+    rw [h, Nat.cast_zero] at hn
+    exact hn rfl
+  have : degree (X ^ n - C x) ≠ 0 := by
+    rw [degree_X_pow_sub_C hn' x]
+    exact ne_of_gt (WithBot.coe_lt_coe.2 hn')
+  by_cases hx : x = 0
+  · exact ⟨0, by rw [hx, pow_eq_zero_iff hn']⟩
+  · obtain ⟨z, hz⟩ := exists_root _ this <| separable_X_pow_sub_C x hn hx
+    use z
+    simp only [eval_C, eval_X, eval_pow, eval_sub, IsRoot.def] at hz
+    exact sub_eq_zero.1 hz
```

Comment:
```suggestion
    simpa [eval_C, eval_X, eval_pow, eval_sub, IsRoot.def, sub_eq_zero] using hz
```
If `simp` gets you to a point where the answer is `exact`, then `simpa` can often be used instead.